esrally/driver/scheduler.py

```diff
@@ -15,29 +15,114 @@
+  on the task's properties.
+* ``before_request`` is invoked by Rally before a request is executed. ``now`` provides the current timestamp in seconds.
+* Similarly, ``after_request`` is invoked by Rally after a response has been received. ``now`` is the current timestamp,
+  ``weight``, ``unit`` and ``metadata`` are passed from the respective runner. For a bulk request, Rally passes e.g. `
+  ``weight=5000, unit="docs"`` or for a search request ``weight=1, unit="ops"``. Note that when a request has finished
+  with an error, the ``weight`` might be zero (depending on the type of error).
+* ``next`` needs to behave identical to simple schedulers.
+
+``before_request`` and ``after_request`` can be used to adjust the target throughput based on feedback from the runner.
+
+If the scheduler also needs access to the parameter source, provide a ``parameter_source`` property. Rally injects the
+task's parameter source into this property.
+"""
+
+
+def scheduler_for(task: esrally.track.Task):
     """
     Creates a scheduler instance
 
-    :param name: The name under which the scheduler is registered.
-    :param params: A dict containing the parameters for this scheduler instance.
+    :param task: The current task for which a scheduler is needed.
     :return: An initialized scheduler instance.
     """
+    logger = logging.getLogger(__name__)
+    if not task.throttled:
+        return Unthrottled()
+
     try:
-        s = __SCHEDULERS[name]
+        scheduler_class = __SCHEDULERS[task.schedule]
     except KeyError:
-        raise exceptions.RallyError("No scheduler available for name [%s]" % name)
-    return s(params)
+        raise exceptions.RallyError(f"No scheduler available for name [{task.schedule}]")
+
+    # for backwards-compatibility - treat existing schedulers as top-level schedulers
+    if is_legacy_scheduler(scheduler_class):
+        logger.warning("Scheduler [%s] implements a deprecated API. Please adapt it.", scheduler_class)
+        return LegacyWrappingScheduler(task, scheduler_class)
+    elif is_simple_scheduler(scheduler_class):
+        logger.debug("Treating [%s] for [%s] as a simple scheduler.", scheduler_class, task)
+        return UnitAwareScheduler(task, scheduler_class)
+    else:
+        logger.debug("Treating [%s] for [%s] as non-simple scheduler.", scheduler_class, task)
+        return scheduler_class(task)
+
+
+def is_legacy_scheduler(scheduler_class):
+    """
+    Determines whether a scheduler is a legacy implementation that gets passed task parameters instead of only the
+    target throughput.
+    """
+    constructor_params = inspect.signature(scheduler_class.__init__).parameters
+    return len(constructor_params) == 2 and "params" in constructor_params
+
+
+def is_simple_scheduler(scheduler_class):
+    """
+    Determines whether a scheduler is a "simple" scheduler, i.e. it doesn't consider feedback from the parameter source.
+    """
```

Comment:
Is this statement accurate as a "regular" scheduler doesn't have to reference the parameter source just `before_request` and `after_request` - this is more feedback from the runner

Reply:
Ah, you're right. I was heavily refactoring this part of the code shortly before creating the PR. That was a leftover. :) I'll change it.